### ISSUE TITLE
Suggestion: A database support

### ISSUE THREAD

**Issue by itays123** [aaronwlee/Attain#31]
I was thinking, what if we could do something like this:
`class MyDatabase implements AttainDatabase {
  /* methods for communicating with the database, such as: */
  async findAllUsers(){
    return await database.query('SELECT * FROM users');
  }
}`

the AttainDatabase interface will contain a connect method.

and then, provide a method with a generic class that implemets AttainDatabase:
`app.database<MyDatabase>()`

and provide a database instance in the handler functions!
that way, you will be able to change your database with just one line of code!

consider refering to [my repo](https://github.com/itays123/sunny-villas) where I did something similar

thoughts?

**Comment by itays123** [comment]
Creating a pull with my work so far, I really got inspired

**Comment by itays123** [comment]
Waiting for your opinions, then I will create a pull with my work so far

**Comment by aaronwlee** [comment]
Thanks for your all contributions!. @itays123 

This sounds wonderful and you're sample code is what I want to! Obviously, I've considered it for a while. However, sadly, there're some critical problems.

As you know, a lot of people are still using the SQL database, but NoSQL users are also getting increased.

If we can support all the databases(all SQL, all NoSQL), then it must be a great idea to implement this DB feature in our boundary. However, if it can't, we should not force people to use a specific database. 

Even if we can make this kind of repository feature, it takes a lot of time and other features might be having a bad quality.

So, I'd like to ask you that how about this [TypeORM](https://github.com/denolib/typeorm) module??

What I want to do is using this module to support the database and serving a template with small guidelines. Also, we can provide some scaffolding functionality via our CLI.

I'd like to hear your opinion too :)

**Comment by itays123** [comment]
My idea is to have a generic type. 

You can use whatever database you want, as long as you use a class for the queries and write a connect() method.

In my repo I used an in-memory db for development purposes and then switched to a PostgreSQL database when I felt like it, with the change of one line of code. If we can do something like that in this project it will be amazing in my opinion.

As for the CLI, I'm thinking this structure would be nice:

Project
- server.tsx
- db
  - ExampleDatabase.ts (in memory db)
- everything else

And add this code to the project:
db/ExampleDatabase.ts
```ts
class ExampleDatabase implements AttainDatabase {
  private users = [{ id: 1 }, { id: 2 }];
  async connect() {
    console.log('connected');
  } 
  get getUsers() {
    return this.users;
  } 
} 
```

Server.tsx:
```tsx
app.database<ExampleDatabase>();
```
controllers/userController.ts
```ts
router.get('/', (req, res, db) => {
  res.json(db.getUsers());
})
```

And then if someone wants to change the database, all they has to do is to provide the same methods in their new database class.

If you want to do something a little bit different and provide annotations to use, it would be really cool as well.
We can have this structure:

Project
- models
  - User.ts
- everything else

Then just use the model as is, without app.database

**Comment by aaronwlee** [comment]
Oh, I see it's a different concept what I imagine!

If the user uses the ORM then actually I can't say it's worth it. 

In the node.js, people who use ORM used to implement it like this way.

create connection
```ts
let database = null;

export function createConnection() {
  database = await somthingDriver(/*...*/)
};

export database;
```

create a model
```ts
class User {
/*...*/
} 
// or 
const User = Schema({
/*...*/
})

export default DatabaseWrapper(User)
// or 
export default database.getRepository(new User)
```

use a repository
```ts
import User from "userModel.ts";

function middleware(req, res) {
  User.findAll()
  /*...*/
}
```

However, I still love your idea and seems really great when the users will not want to use the ORM. Except, the middleware's third parameter is known as next() or something related to the server context. So I think I need to verify it does make sense.

In conclusion, What I mean is I'll not deny you're idea and if you create this feature I'll accept your pull request! :)

**Comment by itays123** [comment]
yeah it's really cool to use modules like mongoose to do it in node.js, however we need to support all databases therefore we can only provide a database instance. I'm just trying to think out of the box and use the pros of TypeScript to make it easier to use databases and play around with them.
didn't notice you have a next() function, the pend() function is way more useful in my opinion.

I will work on this and send a pull soon!

about the parameters in the middleware function:
is ` (req: Request, res: Response, next: Function, db: T) ` okay for you?

Todo:

- [ ] core/application.ts support

- [ ] run some tests

**Comment by itays123** [comment]
I just had the opportunity to look at your code, it's awesome!

One question about [core/proccess.ts](https://github.com/aaronwlee/attain/blob/master/core/process.ts):
I can see you call the callback of the middleware here:

> ```ts
> middleware.callBack
>      ? await middleware.callBack(req, res)
>      : await attainProcedure(req, res, middleware.next);
> ```

(this code is executed twice, in case of a use() middleware and in the case of any other method that isn't param()).

where do you enter the third argument? (the `next()` function).
I don't want to mess things up so I'll need some help here. 
in the callback type ([here](https://github.com/aaronwlee/attain/blob/master/core/types.ts)) I also don't see any next method

Edited: the way you proccess the middlewares into a single array is brilliant! Saved me a lot of trouble generating new routers with a database field each. now I only need a database field on the app class itself!

**Comment by aaronwlee** [comment]
Nono what i mean that third parameter is known as next function or something related to the server context in the general use cases. For example, Expressjs and koajs are using next function as a third param and oak provides server context in one object. What i want to say is non of these server middlewares never delivered DB because it’s not part of the request or response. That’s why I said we need to concern about it.

**Comment by aaronwlee** [comment]
Obviously, the TypeORM module supports all type of database drivers.

Type of ‘(req: Request, res: Response, next: Function, db: T)‘ is perfect for new middleware👌 

And, you can pass the db value into here 👍 
‘await middleware.callBack(req, res, here)’

**Comment by itays123** [comment]
If I pass it this way wouldn't it be (req, res, db, next)?

And I know it's not something you usually do I just think it would be cool reinventing the box a little bit.

**Comment by aaronwlee** [comment]
There's no next in our case.

**Comment by aaronwlee** [comment]
I think we can expand our App class.

```ts
export class App extends Router {
  #serve?: Server;
  #serveTLS?: Server;
  #process?: Promise<void>;
  #processTLS?: Promise<void>;
  // here to add
  #database: Database;
...

  database = (options) => {
    /* you're code here */
  }
...

  #start = async (server: Server) => {
    for await (const srq of server) {
      // pass database value here
      Process(srq, this.middlewares, this.errorMiddlewares, this.database);
    }
  };
```


in process 
``` ts
let database: Database | undefined = null

const Process = async (
  srq: ServerRequest,
  middlewares: MiddlewareProps[],
  errorMiddlewares: ErrorMiddlewareProps[],
  _database: Database | undifined
) => {
  database = _database
....
  // pass database to procedure
  await attainProcedure(req, res, middleware, database);
....
```

in procedure
```ts
...
await middleware.callBack(req, res, database)
...
```

**Comment by itays123** [comment]
Thank You! That's what I'm going to do

**Comment by itays123** [comment]
Closing for now. I opened a pull. BTW I set the db type to any because I didn't manage to do something else without making the whole app class generic.

**Comment by aaronwlee** [comment]
@itays123 Hi again :) I've seen and merged your new feature and I love it. 

I reopened it because I have a new idea for this database feature.
For now, we only allowed an extended AttainDatabase class as a parameter. But I think it would be really great if we support a single promise-based function to make a connection and return the instance of the database connection.

Here is a sample code
```ts
app.database(async () => {
  const connection = await (/* make a new database connection */)
  
  return connection;
})
```
If we'd able to support this single function as well as the class methods, it'd be really handy to manage the databases depends on the user preference.

Also, if this database method is possible to provide without the `await`, it's might be more clear to managing the code.

```ts
// current
await app.database(class);

// new way without await
app.database(function or class);
```
I'd like to suggest it because other methods are provided without `await` such as `app.use() app.get()`. (Unification)

I can implement this new feature right after you confirmed it.

I'd like to hear your opinions :)

**Comment by itays123** [comment]
This is awesome, I will work on it.
It makes the code clearer and I really like the idea.

**Comment by aaronwlee** [comment]
@itays123
acutally I've already done this. 
 
if you don't mind can you verify my changes?

**Comment by itays123** [comment]
Absolutely. Doing it now.

**Comment by aaronwlee** [comment]
here you can check the changes https://github.com/aaronwlee/attain/tree/1.1.0-beta

here is an example code.
```ts
import { App, Router, Request, Response, AttainDatabase } from "./mod.ts";
import { MongoClient, Database } from "https://deno.land/x/mongo@v0.9.2/mod.ts";

class MongoDatabase extends AttainDatabase {
  async connect() {
    console.log("connected")
  }

  async getSome() {
    console.log("get something")
  }
}

async function DB() {
  const client = new MongoClient()
  await client.connectWithUri("mongodb://localhost:27017")
  const database = client.database("test")
  return database;
}

// allow auto inherit mode (auto inherit the types to the middleware)
const app = App.startWith(DB);
const app = App.startWith(MongoDatabase);
// or
const app = new App<MongoDatabase or Database>() // default is any
app.database(DB)
app.database(MongoDatabase)

// this db params will have automatically inherited types from the app<> or startWith method.
app.use((req, res, db) => {

})
```

you can change it whatever you want! you can deny all these things and let me know if you feel something awkward.

**Comment by itays123** [comment]
# this is perfect!
I see you decided to make the App class generic after all, so the code now is clearer and better. 
let me know if you have some other ideas for this feature

**Comment by itays123** [comment]
Reopening because we need to update the documentation. I'm on it.

**Comment by itays123** [comment]
Closing for now. Awesome work!
